### DISCUSSION TITLE
Lack of interoperability between `custom_access_token_hook` and `.getUser()`

### DISCUSSION THREAD

**Discussion by geoffreygarrett** [supabase/supabase#26543] [Feature Requests]
The new `ssr` package has made it clear, as indicated by the warnings:

```
Using the user object as returned from supabase.auth.getSession() or from some supabase.auth.onAuthStateChange() events could be insecure! This value comes directly from the storage medium (usually cookies on the server) and many not be authentic. Use supabase.auth.getUser() instead which authenticates the data by contacting the Supabase Auth server.
```

The custom claims we set via `custom_access_token_hook` are not propagated to the user whatsoever, even through `app_metadata` set via `custom_access_token_hook`. In this case how should we rely on permissions defined in the `access_token` if I can only retrieve it from `.getSession()`? 

Either I'm missing something, or there's no clear established secure method to do what I want to do in SSR & middleware with a user's JWT.

I have tried to resolve this locally, so I apologise if I'm overlooking something. I appreciate any input on whether this warrants a feature request or if this is a case of PEBKAC.

**Comment by cdub615** [comment]
I am also curious about this issue. I spent a decent amount of time wiring up my custom access token hook, but now it seems like that entire approach is not secure.

I've been trying to figure out if I need to undo all that work and just hit the db whenever I need the relevant info about the logged in user, but if there was a way to access these custom claims from the getUser() call instead that would be amazing.

**Comment by KajSzy** [comment]
@thorwebdev following tutorial [Role-based Access Control (RBAC) User Authorization in Next.js](https://www.youtube.com/watch?v=kwoKmi6inAw)
We are getting to the point where to check user roles we are forced to use `getSession()` which is not recommended.
While `getUser` do not returns neither `access_token` nor role info

The only way to properly ensure session and get `access_token` is to call both `getSession` and `getUser` 
And I'm not 100% sure this is proper way
```ts
const {
    data: { session },
  } = await supabase.auth.getSession();
  const {
    data: { user },
  } = await supabase.auth.getUser(session?.access_token);

const jwt = jwtDecode(session.access_token);
```

**Reply by j4w8n** [reply]
You can trust the access_token from getSession as long as you verify it - not just decode it. But until Asymmetric JWTs come out, that can only be safely done on the server-side; not sure where you're doing it here.
